core-processor/src/ext.rs

```diff
@@ -334,6 +334,237 @@ impl<Context> GrowHandler<Context> for LazyGrowHandler {
+        &mut self,
+        packet: &T,
+        delay: u32,
+    ) -> Result<(), FallibleExtError> {
+        // In case of delayed sending from origin message we keep some gas
+        // for it while processing outgoing sending notes so gas for
+        // previously gasless sends should appear to prevent their
+        // invasion for gas for storing delayed message.
+        match (packet.gas_limit(), delay != 0) {
+            // Zero gasfull instant.
+            //
+            // In this case there is nothing to do.
+            (Some(0), false) => Ok(()),
+
+            // Any non-zero gasfull or zero gasfull with delay.
+            //
+            // In case of zero gasfull with delay it's pretty similar to
+            // gasless with delay case.
+            //
+            // In case of any non-zero gasfull we prevent stealing for any
+            // previous gasless-es's thresholds from gas supposed to be
+            // sent with this `packet`.
+            (Some(_), _) => {
+                let prev_gasless_fee = self
+                    .outgoing_gasless
+                    .saturating_mul(self.ext.context.mailbox_threshold);
+                self.reduce_gas(prev_gasless_fee)?;
+                self.outgoing_gasless = 0;
+                Ok(())
+            }
+
+            // Gasless with delay.
+            //
+            // In this case we must give threshold for each uncovered gasless-es
+            // sent, otherwise they will steal gas from this `packet` that was
+            // supposed to pay for delay.
+            //
+            // It doesn't guarantee threshold for itself.
+            (None, true) => {
+                let prev_gasless_fee = self
+                    .outgoing_gasless
+                    .saturating_mul(self.ext.context.mailbox_threshold);
+                self.reduce_gas(prev_gasless_fee)?;
+                self.outgoing_gasless = 1;
+                Ok(())
+            }
+
+            // Gasless instant.
+            //
+            // In this case there is no need to give any thresholds for previous
+            // gasless-es: only counter should be increased.
+            (None, false) => Ok(()),
+        }
+    }
+
+    fn mark_used(&mut self, reservation_id: ReservationId) -> Result<(), ReservationError> {
+        if let Some(
+            GasReservationState::Created { used, .. } | GasReservationState::Exists { used, .. },
+        ) = self.ext.context.gas_reserver.states().get(&reservation_id)
+        {
+            if *used {
+                Err(ReservationError::InvalidReservationId)
+            } else {
+                debug_assert!(
+                    self.reservation_to_mark.is_none(),
+                    "attempt to mark another reservation"
+                );
+                self.reservation_to_mark = Some(reservation_id);
+                Ok(())
+            }
+        } else {
+            Err(ReservationError::InvalidReservationId)
+        }
+    }
```

Comment:
Duplicate of important logic of `GasReserver`. I'd create almost the same common function `check_used`, so `GasReserver::mark_used` will just set flag to `true` and `ExtMutator::mark_reservation_used` will set `reservation_to_mark`